File: 백준/Gold/1941. 소문난 칠공주/소문난 칠공주.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    static char[][] arr;
    static int ans;
    static int[][] dirs = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};

    static void dfs(int n, int start, ArrayList<Integer> lst) {
        if(n == 7) {
            if(isGroup(lst)) ans++;
            return;
        }

        for (int i = start; i < 25; i++) {
            lst.add(n, i);
            dfs(n+1, i+1, lst);
            lst.remove(n);
        }
    }

    static boolean isGroup(ArrayList<Integer> lst) {
        int sCnt = 0;
        for(int num: lst) {
            int x = num/5;
            int y = num%5;
            if(arr[x][y] == 'S') sCnt++;
        }
        if(sCnt < 4) return false;

        Queue<Integer> q = new LinkedList<>();
        boolean[] visited = new boolean[25];
        visited[lst.get(0)] = true;
        q.add(lst.get(0));
        int cnt = 1;

        while(!q.isEmpty()) {
            int now = q.poll();

            int di = now/5;
            int dj = now%5;

            for (int[] dir: dirs) {
                int ni = di+dir[0];
                int nj = dj+dir[1];

                if(ni<0 || ni>=5 || nj<0 || nj>=5) continue;
                int n = ni*5 + nj;
                if(!lst.contains(n) || visited[n]) continue;

                visited[n] = true;
                q.add(n);
                cnt++;
            }
        }

        return cnt == 7;
    }


    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        arr = new char[5][5];

        for (int i = 0; i < 5; i++) {
            String str = br.readLine();
            arr[i] = str.toCharArray();
        }

        dfs(0, 0, new ArrayList<Integer>());

        System.out.println(ans);
    }
}

File: 백준/Gold/1941. 소문난 칠공주/소문난 칠공주.py
from collections import deque

def bound(lst):
    vtmp = [[1]*5 for _ in range(5)]
    for i, j in lst:
        vtmp[i][j] = 0
    
    q = deque()
    q.append(lst[0])
    while q:
        si, sj = q.popleft()
        vtmp[si][sj] = 1
        for di, dj in ((-1,0),(1,0),(0,-1),(0,1)):
            ni,nj = si+di, sj+dj
            if 0<=ni<5 and 0<=nj<5 and not vtmp[ni][nj]:
                q.append((ni,nj))
    
    for i in range(5):
        for j in range(5):
            if vtmp[i][j] == 0:
                return False

    return True

def count(lst):
    y = 0
    for i, j in lst:
        if arr[i][j] == 'Y':
            y += 1
            if y >= 4:
                return False
    return True

def dfs(n, start, lst):
    global cnt
    if n == 7:
        if count(lst) and bound(lst):
            cnt += 1
        return
    for i in range(start, 25):
        if not v[i]:
            v[i] = 1
            dfs(n+1, i+1, lst+[ij[i]])
            v[i] = 0

arr = [list(input()) for _ in range(5)]
ij = [(i,j) for i in range(5) for j in range(5)]
v = [0]*25
cnt = 0
dfs(0, 0, [])

print(cnt)